enhancements/network/primary-user-defined-networks-for-virtualization-workloads.md

```diff
@@ -0,0 +1,638 @@
+(on the VM). Using a label would allow a controller to monitor which workloads
+are attached to an `IPAMClaim`.
+
+OpenShift Virtualization will see the annotation on the VM, and will proceed to
+the `IPAMClaim` owner reference, and template the launcher pod accordingly -
+with the annotation mentioned in the
+[Persisting VM IP addresses during the migration](#persisting-vm-ip-addresses-during-the-migration)
+section.
+
+Once OVN-Kubernetes reconciles the pod which will host the VM, it will first
+validate the request - i.e. are the requested IPs in the configured ranges of
+the UDN network ? If not, the `IPAMClaim` condition will be set accordingly,
+and allocating the OVN annotation for the pod will fail. The network controller
+will keep attempting to reconcile the pod using exponential backoff until it
+possibly succeeds - in the case the user updates the
+`IPAMClaim.Spec.IPRequests` attribute.
+
+If the IPs being requested are in the configured subnet range of the network,
+the OVN-Kuberntes network controller proceeds to allocate the requested IP
+addresses; it might fail in case the requested IP addresses are already
+allocated in the network. If so, the pod will crashloop, and the `IPAMClaim`
+conditions updated to reflect this.
+
+In the case it succeeds, the `IPAMClaim` conditions are updated w/ a success
+condition, and its `IPAMCLaim.Status.IPs` updated accordingly (as happens today)
+for secondary networks.
+
+A variation of this option would be if the user annotated the VM directly with
+the IPs they want to have available on the primary UDN interface - this way,
+OpenShift Virtualization would create the `IPAMClaim` on behalf of the user,
+requesting whose IPs in its `IPAMClaim.Spec.IPRequests` attribute. While this
+would free the user from having to create the `IPAMClaim` object, the API for
+it would be a bit clunky, since we'd have to rely on a comma separated list of
+IPs as the annotation value to support multiple IPs for an interface.
+
+### Using multus default network annotation
+We could use the
+[multus default network annotation](https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/configuration.md#specify-default-cluster-network-in-pod-annotations)
+to pass extra / override attributes for the primary UDN network.
+
+The user would need to define the network-selection-element on the pod, and it
+would need to point to the `IPAMClaim` object, as done for secondary networks.
+
+While this would be a native way to integrate with the rest of the persistent
+IPs feature, it would be a clumsy flow for primary UDNs, since the user would
+be expected to use network selection elements to configure the primary UDN.
+
+Furthermore, we currently not allow primary UDNs to be attached via network 
+selection elements. We would need to break this convention.
+
+This route could be interesting in the future though, since we expect this to
+become the way to request SR-IOV resources for primary UDNs.
```

Comment:
> Maybe I just don't understand how this would look like - I'm assuming the NSE would have to request the attachment of the **default network**. For clarity, could you paste how you envision the NSE for - let's say - a primary UDN requesting an IPAMClaim and SR-IOV device ?
> 
> Assuming I've understood this correctly (and the NSE we need to configure will point to the default network attachment), I think it is weird to put an NSE customizing the **default** network, and for it to really apply / affect the UDN attachment (the IPAMClaim reference, SR-IOV device). After all, you'll need to pass the data corresponding to the default network, but pass the parameters you want to customize for the primary UDN. That, IMHO, doesn't make for a good fit in this particular case for the API. I can argue it's more confusing.

It doesn't look weirder that using the default network name to refer to the primary network status in the network status annotation, as we plan to do, from https://github.com/ovn-org/ovn-kubernetes/pull/4671:
```
[
  {
    "name": "ovn-kubernetes",
    "interface": "eth0",
    "ips": [
      "10.244.1.6",
      "fd00:10:244:2::6"
    ],
    "mac": "0a:58:0a:f4:01:06",
    "default": false,
    "dns": {}
  },
  {
    "name": "ovn-kubernetes",
    "interface": "ovn-udn1",
    "ips": [
      "10.128.0.3"
    ],
    "mac": "0a:58:0a:80:00:03",
    "default": true,
    "dns": {}
  }
]
```

So at this point we have to come to terms that from the perspective of this API, when addressing the default network name, we might actually be addressing the default network for the namespace whatever that might be.

I am not really sure there is a valid argument to say its fine to do it in `network-status` but not in the NSE in terms of API (or UX).

Reply:
> It doesn't look weirder that using the default network name to refer to the primary network status in the network status annotation, as we plan to do, from [ovn-org/ovn-kubernetes#4671](https://github.com/ovn-org/ovn-kubernetes/pull/4671):
> 
> ```
> [
>   {
>     "name": "ovn-kubernetes",
>     "interface": "eth0",
>     "ips": [
>       "10.244.1.6",
>       "fd00:10:244:2::6"
>     ],
>     "mac": "0a:58:0a:f4:01:06",
>     "default": false,
>     "dns": {}
>   },
>   {
>     "name": "ovn-kubernetes",
>     "interface": "ovn-udn1",
>     "ips": [
>       "10.128.0.3"
>     ],
>     "mac": "0a:58:0a:80:00:03",
>     "default": true,
>     "dns": {}
>   }
> ]
> ```
> 

I disagree; to me, it totally does - they are in different elements in the array. The only thing confusing there is the network name; and you can still differentiate using the interface name. 

IMHO that's totally different from using an NSE like this one:
```json
[
    {
        "name":"default",              # relates to a NAD with the default cluster network
        "namespace":"blue-ns",  
        "ipam-claim-reference": "as-we-have.now" # customizes the primary UDN attachment
    }
]
```

In case I am making an error with the NSE contents, please let me know; I've asked for it in https://github.com/openshift/enhancements/pull/1673#discussion_r1756659800, but you probably didn't see it.

> So at this point we have to come to terms that from the perspective of this API, when addressing the default network name, we might actually be addressing the default network for the namespace whatever that might be.

I will give it another go at the interface reporting PR to see if I can actually reply with the network name of the NAD for the primary UDN. If I manage to, I'd say your comment wouldn't apply any more.

> 
> I am not really sure there is a valid argument to say its fine to do it in `network-status` but not in the NSE in terms of API (or UX).

Reply:
> I disagree; to me, it totally does - they are in different elements in the array. The only thing confusing there is the network name; and you can still differentiate using the interface name.

Then we agree we disagree.

That's not the only confusing thing. The fact that `"default": true` is set for the network that is not the default network is also confusing. 

That fact that the array has two elements and therefor its fine it's a tunnel vision specific to your use case, it doesn't mean it is API fine.

Using interface name to distinguish relies on the unspecified convention that `eth0` is the main interface, that wouldn't be good either.

We agreed we had to redefine what `default` means, and maybe the NSE on the default network affecting the primary network fits under that new definition.

However, the NSE might as well be defined addressing the primary UDN network name, nothing would be preventing us being able to handle it as long as OVNK is handling the CNI call at the end, it doesn't really matter if the CNI call is for the primary or the default network, we can be aware and handle it.

Then the network status would be the opposite, and will report twice the primary network. I guess that is fine as well if it is fine the other way around.

Reply:
The NSE could be
```
[
    {
        "name":"ovn-kubernetes",              # relates to a NAD with the default cluster network
        "namespace":"blue-ns",  
        "ipam-claim-reference": "as-we-have.now" # customizes the primary UDN attachment
    }
]
```

I guessed I missed the point that the NAD itself has to exist and that might be inconvenient. But we can address the primary udn nad:
```
[
    {
        "name":"blue-primary-udn",              # relates to a NAD with the default cluster network
        "namespace":"blue-ns",  
        "ipam-claim-reference": "as-we-have.now" # customizes the primary UDN attachment
    }
]
```

Not sure of the format either, I guess not important.

Reply:
Right now we get a request for the default network and we additionally plumb the primary network. Why wouldn't we be able to get a request for the primary network and additionally plumb the default network?

Reply:
> Maybe to keep this discussion short, I think it is feasible to use the NSE and reasonable to highlight the positives of that option in the alternative. That would be my whole ask.

I agree that it is feasible, and have committed to experiment with it already. We're currently PoCing it.

I've also said I would highlight the positives of it in the alternative section.